rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java

```diff
@@ -75,8 +98,190 @@ public NodesLockRestorationManager apply(RMCore rmCore) {
+            if (!im.getDeployingAndLostNodes().isEmpty()) {
+                // if there are deploying nodes, we will not recover
+                rmCore.getDbManager().removeAllNodesFromNodeSource(nodeSourceName);
+                recoverNodes = false;
+            } else {
+                nodeSourceToDeploy.setInfrastructureManager(im);
+            }
+        }
+        return recoverNodes;
+    }
+
+    private boolean existPersistedNodes(String nodeSourceName, boolean nodesRecoverable) {
+        boolean recoverNodes = false;
+        if (PAResourceManagerProperties.RM_NODES_RECOVERY.getValueAsBoolean() && nodesRecoverable) {
+            Collection<RMNodeData> nodesData = rmCore.getDbManager().getNodesByNodeSource(nodeSourceName);
+            if (nodesData.isEmpty()) {
+                logger.info("No node found in database for node source: " + nodeSourceName);
+            } else {
+                recoverNodes = true;
+            }
+        }
+        return recoverNodes;
+    }
+
+    private void recoverDeployedNodeSourceIfNeeded(NodeSourceData nodeSourceDescriptor, String nodeSourceName) {
+        try {
+            logger.info("Recover node source " + nodeSourceName);
+            // retrieve node source status
+            boolean deployNodeSource = false;
+            if (nodeSourceDescriptor.getStatus().equals(NodeSourceStatus.NODES_DEPLOYED)) {
+                deployNodeSource = true;
+            }
+            // reset node source status
+            nodeSourceDescriptor.setStatus(NodeSourceStatus.NODES_UNDEPLOYED);
+            rmCore.createNotActiveNodeSource(nodeSourceDescriptor);
+            // repeat node source status
+            if (deployNodeSource) {
+                rmCore.deployNodeSource(nodeSourceName);
+            }
+        } catch (Throwable t) {
+            logger.error("Failed to recover node source " + nodeSourceName, t);
+            rmCore.removeNodeSource(nodeSourceName);
+            rmCore.getDbManager().removeNodeSource(nodeSourceName);
+        }
+    }
+
+    private void logPersistedNodeSourceInfo(Collection<NodeSourceData> nodeSources) {
+        if (nodeSources.isEmpty()) {
+            logger.info("No node source found in database");
+        } else {
+            if (nodeSources.size() < 10) {
+                logger.info("Node sources found in database: " + Arrays.toString(nodeSources.toArray()));
+            } else {
+                logger.info("Number of node sources found in database: " + nodeSources.size());
+            }
+        }
+    }
+
+    public void recoverNodes(NodeSource nodeSource) {
+        logger.info(START_TO_RECOVER_NODES); // this log line is important for performance tests
+        int lookUpTimeout = PAResourceManagerProperties.RM_NODELOOKUP_TIMEOUT.getValueAsInt();
+        String nodeSourceName = nodeSource.getName();
+
+        makeSureNodeSourceHasNoNode(nodeSource, nodeSourceName);
```

Comment:
Can you make the name a bit more clear like delete nodes or whatever it does. I am not sure what "makeSure"really means. :-D

Reply:
done